File: skills/aox-marketplace/api_server.js
#!/usr/bin/env node
/**
 * AOX Marketplace API Server
 * Serves listings.json with CORS for Vercel frontend
 */

import http from 'http';
import fs from 'fs';
import path from 'path';

const PORT = process.env.MARKETPLACE_API_PORT || 3200;
const LISTINGS_FILE = path.join(process.env.HOME || '/home/ubuntu', '.openclaw/agents/marketplace/output/listings.json');
const ALLOWED_ORIGINS = [
  'https://aox.llc',
  'https://www.aox.llc',
  'https://aox-app.vercel.app',
  'https://aox-ebooks.vercel.app',
  'http://localhost:3000',
  'http://localhost:3001'
];

// CORS headers
const corsHeaders = (origin) => ({
  'Access-Control-Allow-Origin': origin || '*',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type, Authorization',
  'Content-Type': 'application/json'
});

const server = http.createServer((req, res) => {
  const origin = req.headers.origin;
  const allowedOrigin = ALLOWED_ORIGINS.includes(origin) ? origin : ALLOWED_ORIGINS[0];
  
  // Handle preflight
  if (req.method === 'OPTIONS') {
    res.writeHead(204, corsHeaders(allowedOrigin));
    res.end();
    return;
  }
  
  const url = new URL(req.url, `http://${req.headers.host}`);
  
  // Health check
  if (url.pathname === '/health') {
    res.writeHead(200, corsHeaders(allowedOrigin));
    res.end(JSON.stringify({ 
      status: 'ok', 
      service: 'aox-marketplace-api',
      timestamp: new Date().toISOString()
    }));
    return;
  }
  
  // GET /leads - Available listings
  if (url.pathname === '/leads' && req.method === 'GET') {
    try {
      if (!fs.existsSync(LISTINGS_FILE)) {
        res.writeHead(200, corsHeaders(allowedOrigin));
        res.end(JSON.stringify({ listings: [], count: 0 }));
        return;
      }
      
      const data = JSON.parse(fs.readFileSync(LISTINGS_FILE, 'utf8'));
      const listings = data.listings || [];
      
      // Filter by status if requested
      const status = url.searchParams.get('status') || 'available';
      const filtered = status === 'all' 
        ? listings 
        : listings.filter(l => l.status === status);
      
      // Return public data only
      const publicListings = filtered.map(l => ({
        id: l.id,
        status: l.status,
        category: l.category,
        title: l.title,
        description: l.description,
        score: l.score,
        tier: l.tier,
        price: l.price,
        payment_token: l.payment_token,
        listed_at: l.listed_at,
        expires_at: l.expires_at,
        metadata: l.public_metadata || {}
      }));
      
      res.writeHead(200, corsHeaders(allowedOrigin));
      res.end(JSON.stringify({ 
        listings: publicListings, 
        count: publicListings.length,
        updated_at: data.updated_at 
      }));
      
    } catch (error) {
      console.error('Error reading listings:', error);
      res.writeHead(500, corsHeaders(allowedOrigin));
      res.end(JSON.stringify({ error: 'Failed to read listings' }));
    }
    return;
  }
  
  // GET /leads/:id - Specific lead
  if (url.pathname.match(/^\/leads\/[^\/]+$/) && req.method === 'GET') {
    const leadId = url.pathname.split('/')[2];
    
    try {
      const data = JSON.parse(fs.readFileSync(LISTINGS_FILE, 'utf8'));
      const listing = data.listings?.find(l => l.id === leadId);
      
      if (!listing) {
        res.writeHead(404, corsHeaders(allowedOrigin));
        res.end(JSON.stringify({ error: 'Lead not found' }));
        return;
      }
      
      if (listing.status !== 'available') {
        res.writeHead(400, corsHeaders(allowedOrigin));
        res.end(JSON.stringify({ error: 'Lead not available' }));
        return;
      }
      
      res.writeHead(200, corsHeaders(allowedOrigin));
      res.end(JSON.stringify({
        id: listing.id,
        title: listing.title,
        description: listing.description,
        price: listing.price,
        payment_token: listing.payment_token,
        payment_address: listing.payment_address,
        tier: listing.tier,
        score: listing.score
      }));
      
    } catch (error) {
      res.writeHead(500, corsHeaders(allowedOrigin));
      res.end(JSON.stringify({ error: 'Server error' }));
    }
    return;
  }
  
  // POST /webhook/new-lead - Triggered by Policy Gate
  if (url.pathname === '/webhook/new-lead' && req.method === 'POST') {
    let body = '';
    req.on('data', chunk => body += chunk);
    req.on('end', () => {
      try {
        const data = JSON.parse(body);
        console.log('New lead webhook received:', data.lead_id);
        
        // Trigger marketplace agent to process
        // In production: spawn process or queue job
        res.writeHead(200, corsHeaders(allowedOrigin));
        res.end(JSON.stringify({ status: 'received', lead_id: data.lead_id }));
      } catch (e) {
        res.writeHead(400, corsHeaders(allowedOrigin));
        res.end(JSON.stringify({ error: 'Invalid JSON' }));
      }
    });
    return;
  }
  
  // 404
  res.writeHead(404, corsHeaders(allowedOrigin));
  res.end(JSON.stringify({ error: 'Not found' }));
});

server.listen(PORT, '0.0.0.0', () => {
  console.log(`✅ AOX Marketplace API running on port ${PORT}`);
  console.log(`📁 Serving: ${LISTINGS_FILE}`);
  console.log(`🌐 CORS allowed origins:`);
  ALLOWED_ORIGINS.forEach(o => console.log(`   - ${o}`));
  console.log('');
  console.log('Endpoints:');
  console.log(`  GET  http://your-ec2-ip:${PORT}/health`);
  console.log(`  GET  http://your-ec2-ip:${PORT}/leads`);
  console.log(`  GET  http://your-ec2-ip:${PORT}/leads/<id>`);
  console.log(`  POST http://your-ec2-ip:${PORT}/webhook/new-lead`);
});

// Graceful shutdown
process.on('SIGTERM', () => {
  console.log('Shutting down...');
  server.close(() => process.exit(0));
});

process.on('SIGINT', () => {
  console.log('Shutting down...');
  server.close(() => process.exit(0));
});

File: skills/aox-marketplace/api_server.py
#!/usr/bin/env python3
"""
AOX Marketplace API Server
Simple HTTP server for aox-app to fetch listings
"""

import os
import json
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs

LISTINGS_FILE = os.path.expanduser('~/.openclaw/agents/marketplace/output/listings.json')
PORT = int(os.getenv('MARKETPLACE_API_PORT', '8080'))

class MarketplaceHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        # Custom logging
        print(f"[{self.log_date_time_string()}] {args[0]}")
    
    def _send_json(self, data, status=200):
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())
    
    def do_GET(self):
        parsed = urlparse(self.path)
        path = parsed.path
        params = parse_qs(parsed.query)
        
        if path == '/api/leads':
            # Get all available listings
            self._handle_get_listings(params)
        elif path.startswith('/api/leads/'):
            # Get specific lead
            lead_id = path.split('/')[-1]
            self._handle_get_lead(lead_id)
        elif path == '/health':
            self._send_json({'status': 'ok', 'service': 'aox-marketplace'})
        else:
            self._send_json({'error': 'Not found'}, 404)
    
    def do_POST(self):
        parsed = urlparse(self.path)
        path = parsed.path
        
        if path == '/api/purchase':
            self._handle_purchase()
        elif path == '/webhook/payment':
            self._handle_payment_webhook()
        else:
            self._send_json({'error': 'Not found'}, 404)
    
    def _handle_get_listings(self, params):
        """Return available listings."""
        try:
            if not os.path.exists(LISTINGS_FILE):
                self._send_json({'listings': []})
                return
            
            with open(LISTINGS_FILE) as f:
                data = json.load(f)
            
            listings = data.get('listings', [])
            
            # Filter by status
            status_filter = params.get('status', ['available'])[0]
            if status_filter != 'all':
                listings = [l for l in listings if l['status'] == status_filter]
            
            # Filter by category
            category = params.get('category', [None])[0]
            if category:
                listings = [l for l in listings if l['category'].lower() == category.lower()]
            
            # Filter by tier
            tier = params.get('tier', [None])[0]
            if tier:
                listings = [l for l in listings if l['tier'].lower() == tier.lower()]
            
            # Return public data only (no payment addresses, no private hashes)
            public_listings = []
            for listing in listings:
                public_listings.append({
                    'id': listing['id'],
                    'status': listing['status'],
                    'category': listing['category'],
                    'title': listing['title'],
                    'description': listing['description'],
                    'score': listing['score'],
                    'tier': listing['tier'],
                    'price': listing['price'],
                    'payment_token': listing['payment_token'],
                    'listed_at': listing['listed_at'],
                    'expires_at': listing['expires_at'],
                    'metadata': listing.get('public_metadata', {})
                })
            
            self._send_json({'listings': public_listings, 'count': len(public_listings)})
            
        except Exception as e:
            self._send_json({'error': str(e)}, 500)
    
    def _handle_get_lead(self, lead_id: str):
        """Get specific lead (for purchase flow)."""
        try:
            with open(LISTINGS_FILE) as f:
                data = json.load(f)
            
            listings = data.get('listings', [])
            listing = next((l for l in listings if l['id'] == lead_id), None)
            
            if not listing:
                self._send_json({'error': 'Lead not found'}, 404)
                return
            
            if listing['status'] != 'available':
                self._send_json({'error': 'Lead not available'}, 400)
                return
            
            # Return with payment address
            self._send_json({
                'id': listing['id'],
                'title': listing['title'],
                'description': listing['description'],
                'price': listing['price'],
                'payment_token': listing['payment_token'],
                'payment_address': listing.get('payment_address'),
                'tier': listing['tier'],
                'score': listing['score']
            })
            
        except Exception as e:
            self._send_json({'error': str(e)}, 500)
    
    def _handle_purchase(self):
        """Initiate purchase."""
        content_length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(content_length).decode()
        
        try:
            data = json.loads(body)
            lead_id = data.get('lead_id')
            buyer_email = data.get('buyer_email')
            
            # In production: validate, generate unique payment session, etc.
            self._send_json({
                'status': 'pending',
                'lead_id': lead_id,
                'message': 'Send payment to the provided address'
            })
            
        except Exception as e:
            self._send_json({'error': str(e)}, 400)
    
    def _handle_payment_webhook(self):
        """Receive payment notification from blockchain monitoring."""
        content_length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(content_length).decode()
        
        try:
            data = json.loads(body)
            # Process payment confirmation
            # This would trigger the marketplace agent to deliver lead
            self._send_json({'status': 'received'})
        except Exception as e:
            self._send_json({'error': str(e)}, 400)

def run_server():
    server = HTTPServer(('0.0.0.0', PORT), MarketplaceHandler)
    print(f"AOX Marketplace API running on port {PORT}")
    print(f"Endpoints:")
    print(f"  GET  http://localhost:{PORT}/api/leads")
    print(f"  GET  http://localhost:{PORT}/api/leads/<id>")
    print(f"  POST http://localhost:{PORT}/api/purchase")
    print(f"  GET  http://localhost:{PORT}/health")
    server.serve_forever()

if __name__ == '__main__':
    run_server()
